### ISSUE TITLE
Arcade Collision Detection

### ISSUE THREAD

**Issue by Jack0815** [photonstorm/phaser#1424]
I just updated to 2.2.1 and was looking forward to new features and fixes. But with the new update I get a lot of trouble with the collision detection.
An example is just the breakout-example of phaser itself. When I play it for quite some time (besides the fact that the paddle is flickering aournd some times) some of the collisions are not recognized and the ball just flies through the bricks. I have similar issues with 2 other games. It happens often, if you have hits directly followed by each other (e.g. between two lines of bricks in breakout).

Setup:
Phaser 2.2.1
Tested on:
Desktop Browser: Firefox (up-to-date)
Cordova-Build: Nexus 5 (up-to-date)

EDIT: After turning my game into SlowMotion (to see the debug bounding boxes) it did not happen so far. BUT it did also not happen with earlier Phaser versions.

**Comment by pnstickne** [comment]
Notes about having the update transform done by the Physics:

- updateTransform is _relatively_ cheap (as in less than 0.1%), at least according to my desktop analysis.
- this change needs to be applied for all Physic types, not just Arcade
- this would result in the current odd behavior is the user moved the sprite (not via Physics) in an update

I'm a fan of calling it after every postUpdate (or "logic update"), but allowing it to be skipped in the render ("render update") _if_ there was a postUpdate ("logic update") for that game loop cycle.

**Comment by photonstorm** [comment]
The more I think about this the more I think it needs decoupling from physics or even Sprite.postUpdate at all, and just called on a global stage level basis every single logic update (in the core game loop). This would propagate the entire display list.

The reasoning is that while it makes sense that a Sprite should do it in its own postUpdate, this breaks if the Sprite is part of a Group that is itself being modified. For example a Group being tweened (ala a space invaders group of aliens) and yet each alien is needing its own physics updates applied too. If we only update the aliens in their own postUpdate then their coordinates will be wrong because the parent wasn't taken into consideration.

This actually simplifies everything - just calling it in the core game loop and should cover all bases fully. It's also the easiest to implement.

**Comment by pnstickne** [comment]
Well, there is my proposal. It works with the failing cases report and the assumption / rationale is outline in the PR.

**Comment by pjbaron** [comment]
Coming late (and poorly informed, and a little drunk) to this discussion (caveat emptor):
- iirc updateTransform is local to the display object only, it doesn't traverse the parent tree
- groups process individuals within themselves for various update stuff that might change their transform matrix
- groups can be held within other groups (?) potentially causing complex and hard to trace update chains
If I'm correct in these three things then the most robust way to deal with the transforms is to process the entire display list, once per 'logic update', and in a single function call - basically what Rich said.  Anything else seems guaranteed to fail given appropriate circumstances (e.g. parent manipulation by children, especially in call-backs outside the usual update loop).
If the transforms can be changed by the render logic too, then the whole tree starting from any such altered parent needs to be updated again to ensure correct behaviour and display... but I can't think of a valid justification for render logic to adjust the transforms so perhaps it doesn't.

I haven't had a chance to look at your new commit pnstickne (I'm still in Christmas vacation mode) but I've been following this thread and believe that there's really only one or two genuinely robust solutions to this problem so I'm throwing this out there for discussion in case you haven't already covered this ground :)

**Comment by pnstickne** [comment]
@pjbaron That is my latest / revised assumption and the recent change does just that - every "logic update" starts a worldTransform from the stage, which iterates the scene graph.

I've outline the exact assumption in PR.

---

Previously I asserted / hinted that the WT _might_ also be done before a render, but I've since revised that with the assumption that only updates can "move" things. This assumption/rule simplifies the logic and allows skipping a WT under render catch up situations (where there are no update logic calls).

There should probably be an exposed function to allow the user to "`syncWorld`" if user/game code moves objects *outside* of the update phase - which would be in response to a direct browser event or `setTimeout`, etc.

I know of a few _fringe_ cases where this occurs in Phaser (ie. onFullScreenChange with click trampolines or onFullScreenError), but I believe that the input system - which is the main concern for such external events - moves all the relevant callbacks into the update phase. (It would be prudent to review this though and document any such cases that do not.)
